lib/web_ui/lib/src/engine/html/shaders/shader.dart

```diff
@@ -180,29 +167,174 @@ class GradientLinear extends EngineGradient {
       final double fromY = matrix4.transformedY + centerY;
       matrix4.transform(to.dx - centerX, to.dy - centerY);
       gradient = ctx!.createLinearGradient(fromX - offsetX, fromY - offsetY,
-          matrix4.transformedX + centerX - offsetX, matrix4.transformedY - offsetY + centerY);
+          matrix4.transformedX + centerX - offsetX,
+          matrix4.transformedY - offsetY + centerY);
     } else {
-      gradient = ctx!.createLinearGradient(from.dx - offsetX, from.dy - offsetY, to.dx - offsetX, to.dy - offsetY);
+      gradient = ctx!.createLinearGradient(
+          from.dx - offsetX, from.dy - offsetY, to.dx - offsetX,
+          to.dy - offsetY);
     }
+    _addColorStopsToCanvasGradient(gradient, colors, colorStops);
+    return gradient;
+  }
+
+  /// Creates a linear gradient with tiling repeat or mirror.
+  html.CanvasPattern _createGlGradient(html.CanvasRenderingContext2D? ctx,
+      ui.Rect? shaderBounds, double density) {
+    assert(shaderBounds != null);
+    int widthInPixels = shaderBounds!.width.ceil();
+    int heightInPixels = shaderBounds.height.ceil();
+    assert(widthInPixels > 0 && heightInPixels > 0);
+
+    // Render gradient into a bitmap and create a canvas pattern.
+    _OffScreenCanvas offScreenCanvas =
+    _OffScreenCanvas(widthInPixels, heightInPixels);
+    _GlContext gl = _OffScreenCanvas.supported
+        ? _GlContext.fromOffscreenCanvas(offScreenCanvas._canvas!)
+        : _GlContext.fromCanvas(offScreenCanvas._glCanvas!,
+        webGLVersion == WebGLVersion.webgl1);
+    gl.setViewportSize(widthInPixels, heightInPixels);
+
+    NormalizedGradient normalizedGradient = NormalizedGradient(
+        colors, stops: colorStops);
 
-    final List<double>? colorStops = this.colorStops;
-    if (colorStops == null) {
-      assert(colors.length == 2);
-      gradient.addColorStop(0, colorToCssString(colors[0])!);
-      gradient.addColorStop(1, colorToCssString(colors[1])!);
-      return gradient;
+    _GlProgram glProgram = gl.useAndCacheProgram(
+        _WebGlRenderer.writeBaseVertexShader(),
+        _createLinearFragmentShader(normalizedGradient, tileMode))!;
```

Comment:
I think we can make `useAndCacheProgram` return a non-null value instead of null-asserting at call sites.

Reply:
Done.